stores/slabbuffer.go

```diff
@@ -0,0 +1,505 @@
+		if buffer.Complete {
+			mgr.completeBuffers[gid] = append(mgr.completeBuffers[gid], sb)
+		} else {
+			mgr.incompleteBuffers[gid] = append(mgr.completeBuffers[gid], sb)
+		}
+		mgr.buffersByKey[sb.slabKey.String()] = sb
+	}
+	return mgr, nil
+}
+
+func bufferGID(minShards, totalShards uint8, contractSet uint32) bufferGroupID {
+	var bgid bufferGroupID
+	bgid[0] = minShards
+	bgid[1] = totalShards
+	binary.LittleEndian.PutUint32(bgid[2:], contractSet)
+	return bgid
+}
+
+func (mgr *SlabBufferManager) Close() error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	for _, buffers := range mgr.buffersByKey {
+		if err := buffers.file.Close(); err != nil {
+			return err
+		}
+	}
+	mgr.buffersByKey = nil
+	mgr.incompleteBuffers = nil
+	mgr.completeBuffers = nil
+	return nil
+}
+
+func (mgr *SlabBufferManager) AddPartialSlab(ctx context.Context, data []byte, minShards, totalShards uint8, contractSet uint) ([]object.PartialSlab, error) {
+	gid := bufferGID(minShards, totalShards, uint32(contractSet))
+
+	// Sanity check input.
+	slabSize := bufferedSlabSize(minShards)
+	if minShards == 0 || totalShards == 0 || minShards > totalShards {
+		return nil, fmt.Errorf("invalid shard configuration: minShards=%v, totalShards=%v", minShards, totalShards)
+	} else if contractSet == 0 {
+		return nil, fmt.Errorf("contract set must be set")
+	} else if len(data) > slabSize {
+		return nil, fmt.Errorf("data size %v exceeds size of a slab %v", len(data), slabSize)
+	}
+
+	// Deep copy available buffers. We don't want to block the manager while we
+	// perform disk I/O.
+	mgr.mu.Lock()
+	buffers := append([]*SlabBuffer{}, mgr.incompleteBuffers[gid]...)
+	mgr.mu.Unlock()
+
+	// Find a buffer to use. We use at most 1 existing buffer + 1 new buffer to
+	// avoid splitting the data over too many slabs.
+	var slab object.PartialSlab
+	var slabs []object.PartialSlab
+	var err error
+	var usedBuffers []*SlabBuffer
+	for _, buffer := range buffers {
+		var used bool
+		slab, data, used, err = buffer.recordAppend(data)
+		if err != nil {
+			return nil, err
+		}
+		if used {
+			usedBuffers = append(usedBuffers, buffer)
+			slabs = append(slabs, slab)
+		}
+		if len(usedBuffers) == 1 {
+			break
+		}
+	}
+
+	// If there is still data left, create a new buffer.
+	if len(data) > 0 {
+		var sb *SlabBuffer
+		err = mgr.s.retryTransaction(func(tx *gorm.DB) error {
+			sb, err = createSlabBuffer(mgr.s.db, contractSet, mgr.dir, minShards, totalShards)
+			return err
+		})
+		if err != nil {
+			return nil, err
+		}
+		slab, data, _, err = sb.recordAppend(data)
+		if err != nil {
+			return nil, err
+		}
+		if len(data) > 0 {
+			panic("remaining data after creating new buffer")
+		}
+		usedBuffers = append(usedBuffers, sb)
+		slabs = append(slabs, slab)
+
+		// Add new buffer to the list of incomplete buffers.
+		mgr.mu.Lock()
+		mgr.incompleteBuffers[gid] = append(mgr.incompleteBuffers[gid], sb)
```

Comment:
should we re-sort the buffers by remaining space here?

Reply:
They are never sorted in the first place?